File: android/app/src/main/java/com/tacticalmaps/billing/VoucherManager.kt
package com.tacticalmaps.billing

import android.content.Context
import com.google.android.gms.auth.blockstore.Blockstore
import com.google.android.gms.auth.blockstore.RetrieveBytesRequest
import com.google.android.gms.auth.blockstore.StoreBytesData
import java.security.MessageDigest

/**
 * Offline voucher unlock — Android twin of the iOS VoucherManager.
 * Same salt + SHA-256 hash list, so one set of codes works on both platforms.
 * Redemption persists in prefs AND Block Store, so it survives reinstall.
 */
class VoucherManager(context: Context) {

    private val appContext = context.applicationContext
    private val prefs =
        appContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    var isRedeemed: Boolean
        get() = prefs.getBoolean(KEY_REDEEMED, false)
        private set(value) = prefs.edit().putBoolean(KEY_REDEEMED, value).apply()

    init {
        // Re-seed redemption from Block Store after a reinstall.
        if (!isRedeemed) {
            val req = RetrieveBytesRequest.Builder()
                .setKeys(listOf(KEY_REDEEMED)).build()
            Blockstore.getClient(appContext).retrieveBytes(req)
                .addOnSuccessListener { result ->
                    val v = result.blockstoreDataMap[KEY_REDEEMED]
                        ?.bytes?.toString(Charsets.UTF_8)
                    if (v == "1") isRedeemed = true
                }
        }
    }

    /** Attempt to redeem. Returns true (and persists) on a valid code. */
    fun redeem(code: String): Boolean {
        val normalized = normalize(code)
        if (normalized.isEmpty()) return false
        val hex = sha256Hex(SALT + normalized)
        if (hex !in VALID_HASHES) return false
        isRedeemed = true
        val data = StoreBytesData.Builder()
            .setKey(KEY_REDEEMED)
            .setBytes("1".toByteArray(Charsets.UTF_8))
            .setShouldBackupToCloud(false)
            .build()
        Blockstore.getClient(appContext).storeBytes(data) // best-effort
        return true
    }

    private fun normalize(code: String): String =
        code.uppercase().filter { it.isLetterOrDigit() }

    private fun sha256Hex(s: String): String =
        MessageDigest.getInstance("SHA-256")
            .digest(s.toByteArray(Charsets.UTF_8))
            .joinToString("") { "%02x".format(it) }

    companion object {
        /** Must match SALT in scripts/generate_vouchers.py and iOS. */
        private const val SALT = "tacmap-voucher-v1"

        /**
         * SHA-256(SALT + normalizedCode), lowercase hex.
         * REPLACE with the output of scripts/generate_vouchers.py
         * (same list as iOS VoucherManager.validHashes).
         */
        private val VALID_HASHES: Set<String> = setOf(
            // "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
        )

        private const val PREFS = "entitlement"
        private const val KEY_REDEEMED = "voucher_redeemed"
    }
}

File: ios/TacticalMaps/Billing/VoucherManager.swift
import Foundation
import CryptoKit

/// Offline voucher unlock. Codes are validated against an embedded list of
/// SHA-256 hashes (salted), so the plaintext codes never ship in the binary
/// and can't be recovered by dumping strings from the app.
///
/// Generate codes + hashes with `scripts/generate_vouchers.py` and paste the
/// hash list into `validHashes` below (same list goes into the Android
/// VoucherManager so any code works on both platforms).
///
/// Redemption is stored in the Keychain so it survives reinstall, like the
/// trial clock. Note this is *device-scoped* — a voucher redeemed on an
/// iPhone won't carry to the user's iPad (unlike an App Store purchase).
@MainActor
final class VoucherManager: ObservableObject {

    /// Must match SALT in scripts/generate_vouchers.py.
    private static let salt = "tacmap-voucher-v1"

    /// SHA-256(salt + normalizedCode), lowercase hex.
    /// REPLACE with the output of scripts/generate_vouchers.py.
    private static let validHashes: Set<String> = [
        // "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    ]

    private static let kcRedeemed = "voucherRedeemed"

    @Published private(set) var isRedeemed: Bool

    init() {
        isRedeemed = KeychainStore.bool(for: Self.kcRedeemed)
    }

    /// Uppercases and strips spaces/hyphens so "tacmap 1234..." and
    /// "TACMAP-1234-..." both validate.
    static func normalize(_ code: String) -> String {
        code.uppercased().filter { $0.isLetter || $0.isNumber }
    }

    /// Attempt to redeem. Returns true (and persists) on a valid code.
    @discardableResult
    func redeem(_ code: String) -> Bool {
        let normalized = Self.normalize(code)
        guard !normalized.isEmpty else { return false }
        let digest = SHA256.hash(data: Data((Self.salt + normalized).utf8))
        let hex = digest.map { String(format: "%02x", $0) }.joined()
        guard Self.validHashes.contains(hex) else { return false }
        KeychainStore.set(true, for: Self.kcRedeemed)
        isRedeemed = true
        return true
    }
}
